src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/X509Certificates/UnixPkcs12Reader.cs

```diff
@@ -39,19 +39,13 @@ protected void ParsePkcs12(ReadOnlySpan<byte> data)
 
                 // Windows compatibility: Ignore trailing data.
                 ReadOnlySpan<byte> encodedData = reader.PeekEncodedValue();
+                byte[] dataWithoutTrailing = GC.AllocateUninitializedArray<byte>(encodedData.Length, pinned: true);
```

Comment:
FWIW, I have the same concerns here as I did with the other I reviewed. The pinned object heap is not designed for short-lived objects.

Reply:
+1 This is replacing memory leak with a perf bug that can be worse than the memory leak.

Can you wrap the unmanaged memory in a safe handle and AddRef/Release the safe handle around accesses?

Reply:
I'm not a perf expert but this is not a hot path by any means - I suppose all operations done later with cert are way heavier than any cost of this specific allocation/freeing - note there is also lots of parsing going on even within constructor but that's without doing any actual crypto related stuff.

I honestly think we should either change code so that pinning is not needed at all or invest in changes which make this construct fast (I'd imagine passing stuff to native code isn't as uncommon).

I personally greatly prefer simplicity to small perf gains but if there is evidence this is as bad then maybe it is worth it.

When I run this in a tight loop (that was just ctor not E2E) with different cert sizes OOM was very visible while this potential perf issue wasn't particularly noticable - I'm having a really hard time seeing how it can be worse.

Reply:
This isn't about the perf of this code. It's about the effect it has on the rest of the application.

Reply:
> It's about the effect it has on the rest of the application.

Just to demonstrate what I'm talking about, here it's not the Mean that I care most about (though that it is significantly slower and reflective of the impact this has on not only this code but other code in the process as well), but more the fact that in the non-pinned version there are 0 gen1 and 0 gen2 GCs, and that's no longer true in the pinned version: the POH gets collected as part of gen2, so short-lived allocations that would otherwise be gen0 become gen2, while also leading to excessive fragmentation of the POH, and that affects everything else in the app.

| Method    | Length | Mean         | Error      | StdDev     | Ratio | RatioSD | Gen0   | Gen1   | Gen2   |
|---------- |------- |-------------:|-----------:|-----------:|------:|--------:|-------:|-------:|-------:|
| NotPinned | 10     |     7.054 ns |  0.0653 ns |  0.0545 ns |  1.00 |    0.00 | 0.0004 |      - |      - |
| Pinned    | 10     |   234.552 ns |  4.7442 ns |  4.6595 ns | 33.26 |    0.67 | 0.0014 | 0.0014 | 0.0014 |
|           |        |              |            |            |       |         |        |        |        |
| NotPinned | 1000   |    78.747 ns |  0.7053 ns |  0.6252 ns |  1.00 |    0.00 | 0.0111 |      - |      - |
| Pinned    | 1000   | 1,175.699 ns | 23.2682 ns | 31.8498 ns | 15.03 |    0.41 | 0.0267 | 0.0267 | 0.0267 |

```C#
using BenchmarkDotNet.Attributes;
using BenchmarkDotNet.Running;

BenchmarkSwitcher.FromAssembly(typeof(Tests).Assembly).Run(args);

[MemoryDiagnoser]
public class Tests
{
    [Params(10, 1000)]
    public int Length { get; set; }

    [Benchmark(Baseline = true)]
    public byte[] NotPinned() => GC.AllocateUninitializedArray<byte>(Length, pinned: false);

    [Benchmark]
    public byte[] Pinned() => GC.AllocateUninitializedArray<byte>(Length, pinned: true);
}
```

Reply:
Out of curiosity - does allocating unpinned and then using scoped `fixed` statement make any difference? Why aren't pinned objects (or at least arrays) implemented in terms of GlobalHAlloc?

Reply:
> Out of curiosity - does allocating unpinned and then using scoped fixed statement make any difference? 

Yes. However, I am not sure whether it is the right choice for this code. I assume that it allocates pinned memory to prevent GC from creating copies of the certificate in process memory when it compacts the heap. Comment that explains why this allocates pinned memory would be nice.

> Why aren't pinned objects (or at least arrays) implemented in terms of GlobalHAlloc?

It would not be possible. GlobalHAlloc allocations are not managed by the GC. You have to free the allocation explicitly. Pinned object heap allocations are managed by the GC. They are collected when no longer referenced, there is no explicit free operation for you to do.

Reply:
Ideally, the pinned object heap would not have these perf characteristics and its performance characteristics would be more like GlobalHAlloc. There is nothing fundamental preventing that. It is not how it behaves today.

Reply:
> I assume that it allocates pinned memory to prevent GC from creating copies of the certificate in process memory when it compacts the heap.

Yep.  Well, the concern is with the private key material instead of the certificates per se, but the point is to avoid GC compaction clones.

Reply:
@stephentoub @jkotas @bartonjs I've updated the code back to the old version and I'm using SafeHandle now, please review